inc/namespace.php

```diff
@@ -503,6 +517,83 @@ function load_plugins() {
 	}
 }
 
+/**
+ * Ensure Tachyon URL is using the current site hostname.
+ *
+ * @param string $tachyon_url The current tachyon URL.
+ * @return string The updated Tachyon URL.
+ */
+function set_tachyon_hostname( string $tachyon_url ) : string {
+	$tachyon_host = wp_parse_url( $tachyon_url, PHP_URL_HOST );
+	$current_host = wp_parse_url( site_url(), PHP_URL_HOST );
+
+	if ( ! $tachyon_host ) {
+		trigger_error( sprintf( 'Error parsing Tachyon URL: %s', esc_url_raw( $tachyon_url ) ), E_USER_WARNING );
+		return $tachyon_url;
+	}
+
+	if ( ! $current_host ) {
+		trigger_error( sprintf( 'Error parsing current site URL: %s', esc_url_raw( site_url() ) ), E_USER_WARNING );
+		return $tachyon_url;
+	}
+
+	// Only do the replacement if the host name is not a subdomain of the Tachyon host.
+	if ( substr( $current_host, -strlen( $tachyon_host ) ) !== $tachyon_host ) {
+		return str_replace( $tachyon_host, $current_host, $tachyon_url );
+	}
+
+	return $tachyon_url;
+}
+
+/**
+ * Ensure the S3 Uploads Bucket URL matches the current site hostname.
+ *
+ * @param array $dirs Uploads directories array.
+ * @return array
+ */
+function set_s3_uploads_bucket_url_hostname( array $dirs ) : array {
+	$s3_uploads = S3_Uploads::get_instance();
+
+	$primary_host = wp_parse_url( get_main_site_url(), PHP_URL_HOST );
+	$s3_host = wp_parse_url( $s3_uploads->get_s3_url(), PHP_URL_HOST );
+	$current_host = wp_parse_url( site_url(), PHP_URL_HOST );
+
+	if ( ! $primary_host ) {
+		trigger_error( sprintf( 'Error parsing main site URL: %s', esc_url_raw( get_main_site_url() ) ), E_USER_WARNING );
+		return $dirs;
+	}
+
+	if ( ! $s3_host ) {
+		trigger_error( sprintf( 'Error parsing S3 bucket URL: %s', esc_url_raw( $s3_uploads->get_s3_url() ) ), E_USER_WARNING );
+		return $dirs;
+	}
+
+	if ( ! $current_host ) {
+		trigger_error( sprintf( 'Error parsing site URL: %s', esc_url_raw( site_url() ) ), E_USER_WARNING );
+		return $dirs;
+	}
+
+	// To support 3rd party CDNs leave the host names as is if the direct
+	// amazonaws.com URL is in use.
+	if ( strpos( $s3_host, '.amazonaws.com' ) !== false ) {
+		return $dirs;
+	}
+
+	// Ensure uploads host at least matches primary site host.
+	if ( $s3_host !== $primary_host ) {
+		$dirs['url'] = str_replace( $s3_host, $primary_host, $dirs['url'] );
+		$dirs['baseurl'] = str_replace( $s3_host, $primary_host, $dirs['baseurl'] );
+	}
+
+	// Only do the replacement if the host name is not a subdomain of the S3 host.
+	if ( substr( $current_host, -strlen( $primary_host ) ) !== $primary_host ) {
```

Comment:
Ditto above operator comment.
```suggestion
	if ( substr( $current_host, -1 * strlen( $primary_host ) ) !== $primary_host ) {
```